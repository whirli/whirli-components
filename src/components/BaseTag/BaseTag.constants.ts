export const PropKeys: Record<string, string> = {
  SIZE: 'size',
  TAG: 'tag',
  THEME: 'theme',
};

export const PropValues: Record<string, string[]> = {
  SIZE: ['default'],
  [PropKeys.TAG]: ['div', 'span', 'button'],
  [PropKeys.THEME]: ['default'],
};
