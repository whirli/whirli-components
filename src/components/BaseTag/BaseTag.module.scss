@import '../../styles/Breakpoints.module.scss';
@import '../../styles/Spacers.scss';

// default styles
.tag {
    display: inline-block;
    position: relative;
    background: none;
    border: none;
    box-shadow: none;
}

.size-default {
    padding: $spacer2Xs $spacerSm;
}

.theme-default {
    background: grey;
    color: black
}
// breakpoint styles
$packageBreakpointStyles: ();