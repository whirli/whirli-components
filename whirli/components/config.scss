@import '@whirli-components/styles/Helpers.scss';

$breakpointPrefix: BREAKPOINT_;

$breakpoints: (
    tablet: (
        prefix: t,
        min-width: 672px,
    ),
    small_desk: (
        prefix: sd,
        min-width: 1056px
    ),
    large_desk: (
        prefix: ld,
        min-width: 1312px
    ),
    huge_desk: (
        prefix: hd,
        min-width: 1584px
    ),
);

:export {
    prefix: $breakpointPrefix;
    #{$breakpointPrefix}default: 0px;
    @each $name, $value in $breakpoints {
        #{$breakpointPrefix}#{$name}: map-deep-get($breakpoints, #{$name}, 'min-width')
    }
}