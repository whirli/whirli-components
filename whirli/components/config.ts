import BaseAside from './BaseAside/BaseAside.config';
import BaseButton from './BaseButton/BaseButton.config';
import BaseColumn from './BaseColumn/BaseColumn.config';
import BaseContainer from './BaseContainer/BaseContainer.config';
import BaseGrid from './BaseGrid/BaseGrid.config';
import BaseImage from './BaseImage/BaseImage.config';
import BaseMain from './BaseMain/BaseMain.config';
import BaseSection from './BaseSection/BaseSection.config';
import BaseSpacer from './BaseSpacer/BaseSpacer.config';
import BaseText from './BaseText/BaseText.config';

export default {
  BaseAside,
  BaseButton,
  BaseColumn,
  BaseContainer,
  BaseGrid,
  BaseImage,
  BaseMain,
  BaseSection,
  BaseSpacer,
  BaseText,
};
