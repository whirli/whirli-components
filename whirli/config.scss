@import '@whirli-components/src/styles/Helpers.scss';

$breakpoints: (
    tablet: (
        prefix: t,
        min-width: 48rem,
    ),
    desktop: (
        prefix: d,
        min-width: 80rem,
    )
);

:export {
    @each $name, $value in $breakpoints {
        BREAKPOINT_#{to-upper-case(#{$name})}: map-deep-get($breakpoints, #{$name}, 'min-width')
    }
}