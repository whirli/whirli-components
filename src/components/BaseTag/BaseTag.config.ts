// @ts-ignore
import { PropType } from 'vue';
import * as PropHelpers from '@whirli-components/helpers/props';
import { ComponentStyles, ComponentProps } from '@whirli-components/@types/components';

// Local config
import config from '@whirli-local/components/BaseTag/BaseTag.config';

// Package config
import { PropKeys, PropValues } from './BaseTag.constants';
import type { PropSize, PropTag, PropTheme } from './BaseTag.types';

export const ConfigStyles: ComponentStyles = {
  ...config.styles,
  [PropKeys.SIZE]: {
    classes: {
      default: 'size-default',
      ...config.styles[PropKeys.SIZE]?.classes,
    } as Record<PropSize, string>,
  },
  [PropKeys.THEME]: {
    classes: {
      default: 'theme-default',
      ...config.styles[PropKeys.THEME]?.classes,
    } as Record<PropTheme, string>,
  },
};

export const ConfigProps: ComponentProps = {
  [PropKeys.SIZE]: {
    required: false,
    type: String as PropType<PropSize>,
    default: 'default',
    ...PropHelpers.stylePropValidation(ConfigStyles, PropKeys.SIZE),
  },
  [PropKeys.TAG]: {
    required: false,
    type: String as PropType<PropTag>,
    validator: (value: PropTag): boolean => PropValues[PropKeys.TAG].includes(value),
    default: 'div',
  },
  [PropKeys.THEME]: {
    required: false,
    type: String as PropType<PropTheme>,
    default: 'default',
    ...PropHelpers.stylePropValidation(ConfigStyles, PropKeys.THEME),
  },
  ...config.props,
};
