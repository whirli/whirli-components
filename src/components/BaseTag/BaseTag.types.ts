import { PropKeys, PropValues } from './BaseTag.constants';

const propSize: string[] = PropValues[PropKeys.SIZE];
const propTag: string[] = PropValues[PropKeys.TAG];
const propTheme: string[] = PropValues[PropKeys.THEME];

export type PropSize = typeof propSize[number];
export type PropTag = typeof propTag[number];
export type PropTheme = typeof propTheme[number];
