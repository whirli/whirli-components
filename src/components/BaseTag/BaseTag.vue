<template>
  <component :is="props.tag" :class="classes"><slot /></component>
</template>

<script setup="props" lang="ts">
// Vue
import { defineProps } from 'vue';

// Styles
import styles from '@whirli-local/components/BaseTag/BaseTag.module.scss';

// Data
import { ConfigStyles, ConfigProps } from './BaseTag.config';

// Types
import { ComponentStyles as ComponentStylesInterface } from '../../@types/components';
import { Props } from '../../@types/props';

const ComponentStyles: ComponentStylesInterface = ConfigStyles;

const props: Props = defineProps(ConfigProps);

// Classes
import useClasses from '../../@use/class';
const { makeClasses } = useClasses();
const classes = [styles.tag, ...makeClasses(ComponentStyles, props, styles)];
</script>
