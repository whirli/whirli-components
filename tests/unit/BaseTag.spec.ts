import { shallowMount } from '@vue/test-utils';

import { propReturnsClass } from '../helpers';

import BaseTag from '@/components/BaseTag/BaseTag.vue';
import { PropKeys } from '@/components/BaseTag/BaseTag.constants';
import { ConfigStyles } from '@/components/BaseTag/BaseTag.config';

let wrapper: any;
const text = 'Hello World';

describe('BaseTag.vue - Single classes', () => {
  beforeEach(() => {
    wrapper = shallowMount(BaseTag, {
      slots: {
        default: text,
      },
      props: {
        tag: 'span',
      },
    });
  });

  it('renders the component', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('displays the correct text', () => {
    expect(wrapper.text()).toMatch(text);
  });

  it('displays the correct element', () => {
    expect(wrapper.find('span').exists()).toBeTruthy();
  });

  it('has correct classes', () => {
    expect(propReturnsClass(wrapper, ConfigStyles, PropKeys.THEME, 'theme-default')).toBe(true);
    expect(propReturnsClass(wrapper, ConfigStyles, PropKeys.SIZE, 'size-default')).toBe(true);
  });
});
