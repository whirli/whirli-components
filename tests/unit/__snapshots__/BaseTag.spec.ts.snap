// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseTag.vue - Single classes renders the component 1`] = `
<span
  class=""
>
  
  Hello World
  
</span>
`;
